cluster-autoscaler/clusterstate/clusterstate.go

```diff
@@ -1090,3 +1115,114 @@ func fakeNode(instance cloudprovider.Instance) *apiv1.Node {
+	cache.Lock()
+	defer cache.Unlock()
+	cache.cloudProviderNodeInstances[nodeGroup.Id()] = cacheEntry
+}
+
+func (cache *CloudProviderNodeInstancesCache) removeCacheEntryLocked(nodeGroup cloudprovider.NodeGroup) {
+	cache.Lock()
+	defer cache.Unlock()
+	delete(cache.cloudProviderNodeInstances, nodeGroup.Id())
+}
+
+func (cache *CloudProviderNodeInstancesCache) getCacheEntryLocked(nodeGroup cloudprovider.NodeGroup) (*cloudProviderNodeInstancesCacheEntry, bool) {
+	cache.Lock()
+	defer cache.Unlock()
+	cacheEntry, found := cache.cloudProviderNodeInstances[nodeGroup.Id()]
+	return cacheEntry, found
+}
+
+func (cache *CloudProviderNodeInstancesCache) removeEntriesForNonExistingNodeGroupsLocked(nodeGroups []cloudprovider.NodeGroup) {
+	cache.Lock()
+	defer cache.Unlock()
+	nodeGroupExists := map[string]bool{}
+	for _, nodeGroup := range nodeGroups {
+		nodeGroupExists[nodeGroup.Id()] = true
+	}
+	for nodeGroupId := range cache.cloudProviderNodeInstances {
+		if !nodeGroupExists[nodeGroupId] {
+			delete(cache.cloudProviderNodeInstances, nodeGroupId)
+		}
+	}
+}
+
+// GetCloudProviderNodeInstances returns cloud provider node instances for all node groups returned by cloud provider.
+func (cache *CloudProviderNodeInstancesCache) GetCloudProviderNodeInstances() (map[string][]cloudprovider.Instance, error) {
+	results := map[string][]cloudprovider.Instance{}
+	for _, nodeGroup := range cache.cloudProvider.NodeGroups() {
+		nodeGroupInstances, err := cache.GetCloudProviderNodeInstancesForNodeGroup(nodeGroup)
+		if err != nil {
+			return nil, err
+		}
+		results[nodeGroup.Id()] = nodeGroupInstances
+	}
+	return results, nil
+}
+
+// GetCloudProviderNodeInstancesForNodeGroup returns cloud provider node instances for the given node group.
+func (cache *CloudProviderNodeInstancesCache) GetCloudProviderNodeInstancesForNodeGroup(nodeGroup cloudprovider.NodeGroup) ([]cloudprovider.Instance, error) {
+	cacheEntry, found := cache.getCacheEntryLocked(nodeGroup)
+	if found {
+		if cacheEntry.isStale() {
+			klog.Warningf("Entry cloudProviderNodeInstances is stale, refresh time is %v", cacheEntry.refreshTime)
+		}
+		klog.V(5).Infof("Get cached cloud provider node instances for %v", nodeGroup.Id())
+		return cacheEntry.instances, nil
+	}
+	klog.V(5).Infof("Cloud provider node instances for %v hasn't been found in cache, fetch from cloud provider", nodeGroup.Id())
+	nodeGroupInstances, err := nodeGroup.Nodes()
+	if err != nil {
+		return nil, err
+	}
+	cache.updateCacheEntryLocked(nodeGroup, &cloudProviderNodeInstancesCacheEntry{nodeGroupInstances, time.Now()})
+	return nodeGroupInstances, nil
+
+}
+
+// InvalidateCacheEntry removes entry for the given node group from cache.
+func (cache *CloudProviderNodeInstancesCache) InvalidateCacheEntry(nodeGroup cloudprovider.NodeGroup) {
+	klog.V(5).Infof("Invalidate entry in cloud provider node instances cache %v", nodeGroup.Id())
+	cache.removeCacheEntryLocked(nodeGroup)
```

Comment:
I am thinking that maybe instead removing entry here we can just make it stale.
Then we can change the refresh algorithm in such way that we are running it more often (say every 10 seconds) and it refreshes only stale entries?

Then if we have many scale-ups in progress we still will not wait for all to be refreshed every loop. But we will prioritize that those are refreshed more often.

I am thinking about the case when we have multiple GPU-enabled nodegroups for which scale-ups take a lot of time (as GPU nodes are slow to come up). Then we can end up with each loop still taking lots of time. 

But maybe this is premature optimization. WDYT?

Reply:
I see following problem with such solution: We need to refresh all instances every x minutes. If there is a lot of instance to be refreshed during one refresh then we might end up waiting longer than 10 seconds for refreshing scaled up node groups. Other node groups can starve scaled up node groups for some time.